bin/helpers/openstack.sh

```diff
@@ -0,0 +1,357 @@
+}
+
+function verify_downloaded_image() {
+  local download_filename=$1
+  local download_url=$2
+  local download_filepath="/tmp/${download_filename}"
+
+  echo "Downloading ${download_url}/${download_filename}.md5 file"
+  curl "${download_url}/${download_filename}.md5" \
+    -o "${download_filepath}.md5" \
+    --connect-timeout 30 \
+    --max-time 30
+
+  # Verify md5sum of downloaded file
+  echo "Checking md5 sum"
+  md5result=$(
+    cd /tmp || exit;
+    md5sum -c "${download_filename}.md5"
+  )
+  if [[ "${md5result}" != *": OK"* ]]; then
+    echo >&2 "Wrong checksum of downloaded image."
+    echo >&2 "Something might have failed on file transfer."
+    echo >&2 "Please try again."
+    exit 1
+  else
+    echo "Checksum of downloaded image OK"
+  fi
+}
+
+function post_image_upload_checks() {
+  local image_name=$1
+  local disc_format=$2
+  local image_url=$3
+  local is_uploaded_disc_format_as_expected
+  local image_id
+  image_id=$(get_image_id "${upload_name}")
+  server_image_disc_format=$(get_server_image_disc_format "${image_id}")
+  is_uploaded_disc_format_as_expected=$(is_uploaded_disc_format_as_expected "${image_name}" "${disc_format}")
+
+  # Verifu uploaded image checksums
+  verify_uploaded_image "${image_name}" "${disc_format}" "${image_url}"
+
+  # Verify the server image is in the expected disc_format
+  if [[ "${is_uploaded_disc_format_as_expected}" == "false" ]]; then
+    echo >&2 "Host image disc format: ${disc_format}"
+    echo >&2 "Server image disc format: ${server_image_disc_format}"
+    echo >&2 "${upload_name} disc format on the server is not as expected"
+    echo >&2 "Delete the server image and retry the kn init process"
+    exit 1
+  fi
+}
+
+function download_image() {
+  local download_filename=$1
+  local download_url=$2
+  local download_filepath="/tmp/${download_filename}"
+
+  echo "Downloading image to local /tmp/"
+  echo -e "\n"
+  curl "${download_url}/${download_filename}" \
+    -o "${download_filepath}" \
+    --connect-timeout 30 \
+    --max-time 1800
+
+  echo "Verifying the md5 checksum of the downloaded image"
+  echo -e "\n"
+  verify_downloaded_image "${download_filename}" "${download_url}"
+}
+
+function maybe_download_image(){
+  local image_name=$1
+  local disc_format=$2
+  local download_url=$3
+  local download_filename="${image_name}.qcow2"
+  local download_filepath="/tmp/${image_name}.qcow2"
+  local is_image_downloaded_already
+  is_image_downloaded_already=$(is_image_downloaded_already "${download_filepath}")
+
+  if [[ "${is_image_downloaded_already}" == "false" ]]; then
+    download_image "${download_filename}" "${download_url}"
+  else
+    echo "${download_filename} is already downloaded"
+    echo "Verifying the downloaded image checksum"
+    echo -e "\n"
+
+    verify_downloaded_image "${download_filename}" "${download_url}"
+  fi
+}
+
+function maybe_convert_image(){
+  local image_name=$1
+  local disc_format=$2
```

Comment:
To avoid confusion with image format on image downloaded change "disk_format" to "glance_disk_format"